File: src/org/koreader/device/rockchip/RK30xxEPDController.java
/**
 * This file was created by unw on 15. 3. 25 as part of
 * https://github.com/unwmun/refreshU
 */

package org.koreader.device.rockchip;

import android.util.Log;
import android.view.View;

import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@SuppressWarnings("unchecked")
public abstract class RK30xxEPDController
{
    enum EINK_MODE
    {
        EPD_AUTO,
        EPD_FULL,
        EPD_A2,
        EPD_PART,
        EPD_FULL_DITHER,

        EPD_RESET,
        EPD_BLACK_WHITE,
        EPD_TEXT,
        EPD_BLOCK,
        EPD_FULL_WIN,

        EPD_OED_PART,
        EPD_DIRECT_PART,
        EPD_DIRECT_A2,
        EPD_STANDBY,
        EPD_POWEROFF
    }

    public static final String LOGGER_NAME = "luajit-launcher";

    public static final String EPD_NULL = "EPD_NULL";

    public static final String EPD_AUTO = "EPD_AUTO";
    public static final String EPD_FULL = "EPD_FULL";
    public static final String EPD_A2 = "EPD_A2";
    public static final String EPD_PART = "EPD_PART";
    public static final String EPD_FULL_DITHER = "EPD_FULL_DITHER";

    public static final String EPD_RESET = "EPD_RESET";
    public static final String EPD_BLACK_WHITE = "EPD_BLACK_WHITE";
    public static final String EPD_TEXT = "EPD_TEXT";
    public static final String EPD_BLOCK = "EPD_BLOCK";
    public static final String EPD_FULL_WIN = "EPD_FULL_WIN";

    public static final String EPD_OED_PART = "EPD_OED_PART";
    public static final String EPD_DIRECT_PART = "EPD_DIRECT_PART";
    public static final String EPD_DIRECT_A2 = "EPD_DIRECT_A2";
    public static final String EPD_STANDBY = "EPD_STANDBY";
    public static final String EPD_POWEROFF = "EPD_POWEROFF";

    protected static Class<Enum> eInkEnum;
    protected static Method requestEpdModeMethod1;
    protected static Method requestEpdModeMethod2;

    protected static Field isInA2;

    static {
        try {
            eInkEnum = (Class<Enum>) Class.forName("android.view.View$EINK_MODE");
            requestEpdModeMethod1 = View.class.getMethod("requestEpdMode", eInkEnum);
            requestEpdModeMethod2 = View.class.getMethod("requestEpdMode", eInkEnum, boolean.class);
            isInA2 = View.class.getDeclaredField("mIsInA2");
            isInA2.setAccessible(true);

        } catch (ClassNotFoundException e) {
            Log.e(LOGGER_NAME, e.toString());
        } catch (NoSuchMethodException e) {
            Log.e(LOGGER_NAME, e.toString());
        } catch (NoSuchFieldException e) {
            Log.e(LOGGER_NAME, e.toString());
        }
    }

    public static boolean isInA2(View view)
    {
        try {
            boolean value = (Boolean)isInA2.get(view);
            Log.d(LOGGER_NAME, "isInA2 : " + value);
            return value;
        } catch (IllegalAccessException e) {
            Log.e(LOGGER_NAME, e.toString());
            return false;
        }
    }

    public static boolean requestEpdMode(View view, String mode) {

        try {
            requestEpdModeMethod1.invoke(view, stringToEnum(mode));
            return true;
        } catch (IllegalAccessException e) {
            Log.e(LOGGER_NAME, e.toString());
            return false;
        } catch (InvocationTargetException e) {
            Log.e(LOGGER_NAME, e.toString());
            return false;
        }
    }

    public static boolean requestEpdMode(View view, String mode, boolean flag) {
        try {
            requestEpdModeMethod2.invoke(view, stringToEnum(mode), flag);
            return true;
        } catch (IllegalAccessException e) {
            Log.e(LOGGER_NAME, e.toString());
            return false;
        } catch (InvocationTargetException e) {
            Log.e(LOGGER_NAME, e.toString());
            return false;
        }
    }

    public static Object stringToEnum(String str)
    {
        return Enum.valueOf(eInkEnum, str);
    }
}

File: app/src/main/java/org/koreader/launcher/device/epd/rockchip/RK30xxEPDController.kt
/* generic EPD Controller for Android,
 *
 * interface for Boyue T61/T62 clones
 * based on https://github.com/unwmun/refreshU
 *
 * Thanks to @unwmun
 *
 * val EPD_NULL = "EPD_NULL"
 * val EPD_AUTO = "EPD_AUTO"
 * val EPD_FULL = "EPD_FULL"
 * val EPD_A2 = "EPD_A2"
 * val EPD_PART = "EPD_PART"
 * val EPD_FULL_DITHER = "EPD_FULL_DITHER"
 * val EPD_RESET = "EPD_RESET"
 * val EPD_BLACK_WHITE = "EPD_BLACK_WHITE"
 * val EPD_TEXT = "EPD_TEXT"
 * val EPD_BLOCK = "EPD_BLOCK"
 * val EPD_FULL_WIN = "EPD_FULL_WIN"
 * val EPD_OED_PART = "EPD_OED_PART"
 * val EPD_DIRECT_PART = "EPD_DIRECT_PART"
 * val EPD_DIRECT_A2 = "EPD_DIRECT_A2"
 * val EPD_STANDBY = "EPD_STANDBY"
 * val EPD_POWEROFF = "EPD_POWEROFF"
 */

package org.koreader.launcher.device.epd.rockchip

import android.util.Log
import android.view.View
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

abstract class RK30xxEPDController {
    companion object {
        private const val TAG = "EPD"
        private var eInkEnum: Class<Enum<*>>? = null
        private var updateEpdMethod: Method? = null

        init {
            try {
                @Suppress("UNCHECKED_CAST")
                eInkEnum = Class.forName("android.view.View\$EINK_MODE") as Class<Enum<*>>
                updateEpdMethod = View::class.java.getMethod("requestEpdMode",
                    eInkEnum, Boolean::class.javaPrimitiveType)
            } catch (e: ClassNotFoundException) {
                Log.e(TAG, e.toString())
            } catch (e: NoSuchMethodException) {
                Log.e(TAG, e.toString())
            }
        }

        fun requestEpdMode(view: View, mode: String, flag: Boolean): Boolean {
            return try {
                updateEpdMethod!!.invoke(view, stringToEnum(mode), flag)
                true
            } catch (e: IllegalAccessException) {
                Log.e(TAG, e.toString())
                false
            } catch (e: InvocationTargetException) {
                Log.e(TAG, e.toString())
                false
            }
        }

        private fun stringToEnum(str: String): Any {
            val values = eInkEnum!!.enumConstants as Array<Enum<*>>
            return values.first { it.name == str}
        }
    }
}
